library/src/main/java/com/chuckerteam/chucker/api/ChuckerInterceptor.kt

```diff
@@ -80,52 +101,40 @@ class ChuckerInterceptor @JvmOverloads constructor(
                 transaction.isResponseBodyPlainText = false
             }
         }
+    }
 
-        collector.onRequestSent(transaction)
-
-        val startNs = System.nanoTime()
-        val response: Response
-        try {
-            response = chain.proceed(request)
-        } catch (e: IOException) {
-            transaction.error = e.toString()
-            collector.onResponseReceived(transaction)
-            throw e
-        }
-
-        val tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs)
-
+    /**
+     * Processes a [Response] and populates corresponding fields of a [HttpTransaction].
+     */
+    private fun processResponse(response: Response, transaction: HttpTransaction) {
         val responseBody = response.body()
+        val responseEncodingIsSupported = io.bodyHasSupportedEncoding(response.headers().get(CONTENT_ENCODING))
 
-        // includes headers added later in the chain
-        transaction.setRequestHeaders(filterHeaders(response.request().headers()))
         transaction.apply {
-            responseDate = System.currentTimeMillis()
-            this.tookMs = tookMs
+            // includes headers added later in the chain
+            setRequestHeaders(filterHeaders(response.request().headers()))
+            setResponseHeaders(filterHeaders(response.headers()))
+
+            isResponseBodyPlainText = responseEncodingIsSupported
+            requestDate = response.sentRequestAtMillis()
+            responseDate = response.receivedResponseAtMillis()
             protocol = response.protocol().toString()
             responseCode = response.code()
             responseMessage = response.message()
 
             responseContentType = responseBody?.contentType()?.toString()
             responseContentLength = responseBody?.contentLength() ?: 0L
-            setResponseHeaders(filterHeaders(response.headers()))
-        }
 
-        val responseEncodingIsSupported = io.bodyHasSupportedEncoding(response.headers().get(CONTENT_ENCODING))
-        transaction.isResponseBodyPlainText = responseEncodingIsSupported
+            tookMs = (response.receivedResponseAtMillis() - response.sentRequestAtMillis())
```

Comment:
This is definitely a different metric than the one we used before. I don't think we will have major problem now, but it's probably worth to investigate if metrics are consistent with the calculation we used before.

Reply:
My point was that this metric should be more accurate, since it is provided by OkHttp itself.

Reply:
Yup. I've done some testing and we should be fine with the new approach 👍